### PR TITLE
karmada operator: support etcd high availability

### PR DESCRIPTION
**Pull request by calvin0327**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
support etcd high availability, the test result:
we need specify an odd number of replicas for etcd.

image

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by calvin0327** [comment]
@Poor12 @lonelyCZ /cc

File: operator/config/samples/karmada.yaml [hunk 1]
```diff
@@ -9,6 +9,7 @@ spec:
       local:
         imageRepository: registry.k8s.io/etcd
         imageTag: 3.5.3-0
+        replicas: 1
```
Comment (Poor12) [inline operator/config/samples/karmada.yaml:12]: Does the replicas is a required replicas? What's the default value when users do not set it?

File: operator/pkg/init.go [hunk 2]
```diff
@@ -108,6 +108,13 @@ func NewInitJob(opt *InitOptions) *workflow.Job {
 			}
 		}
 
+		if opt.Components.Etcd.Local != nil && opt.Components.Etcd.Local.CommonSettings.Replicas != nil {
```
Comment (Poor12) [inline operator/pkg/init.go:111]: Why replicas must be an odd number?

File: operator/config/samples/karmada.yaml (hunk 1 above)
Comment (calvin0327) [reply operator/config/samples/karmada.yaml:12]: the replicas is required. we should be consistent with other component. if the field .spec is nil, we will be populated default value.  @Poor12

File: operator/pkg/init.go (hunk 2 above)
Comment (calvin0327) [reply operator/pkg/init.go:111]: We recommend the replicas of etcd as a odd number. 
@Poor12 Would it be better to change the error to warning. how do you think?

File: operator/config/samples/karmada.yaml (hunk 1 above)
Comment (Poor12) [reply operator/config/samples/karmada.yaml:12]: ok, I find that default value is 1.

**Comment by Poor12** [comment]
Generally LGTM
/cc @lonelyCZ for comments.

**Comment by lonelyCZ** [comment]
Thanks. I will look it ASAP.

/assign

File: operator/pkg/controlplane/etcd/etcd.go [hunk 3]
```diff
@@ -25,19 +25,31 @@ func EnsureKarmadaEtcd(client clientset.Interface, cfg *operatorv1alpha1.LocalEt
...
 
 func installKarmadaEtcd(client clientset.Interface, name, namespace string, cfg *operatorv1alpha1.LocalEtcd) error {
+	// if the number of etcd is greater than one, we need to concatenate the
+	// peerURL for each member cluster.
+	// memberName is podName generated by etcd statefuleset: ${statefulsetName}-index
+	// memberPeerURL uses the etcd peer headless service name: ${podName}.${namespace}.svc.cluster.local:2380
+	initialClusters := make([]string, *cfg.Replicas)
+	for index := range initialClusters {
+		memberName := fmt.Sprintf("%s-%d", util.KarmadaEtcdName(name), index)
+		memberPeerURL := fmt.Sprintf("%s.%s.svc.cluster.local:%v", memberName, namespace, constants.EtcdListenPeerPort)
```
Comment (lonelyCZ) [inline operator/pkg/controlplane/etcd/etcd.go:35]: There seems an error when starting etcd, could you please take a look? @calvin0327 

```
{"level":"warn","ts":"2023-04-10T06:20:24.484Z","caller":"etcdmain/etcd.go:146","msg":"failed to start etcd","error":"error setting up initial cluster: URL scheme must be http, https, unix, or unixs: karmada-demo-etcd-0.test.svc.cluster.local:2380"}
{"level":"fatal","ts":"2023-04-10T06:20:24.484Z","caller":"etcdmain/etcd.go:204","msg":"discovery failed","error":"error setting up initial cluster: URL scheme must be http, https, unix, or unixs: karmada-demo-etcd-0.test.svc.cluster.local:2380","stacktrace":"go.etcd.io/etcd/server/v3/etcdmain.startEtcdOrProxyV2\n\t/go/src/go.etcd.io/etcd/release/etcd/server/etcdmain/etcd.go:204\ngo.etcd.io/etcd/server/v3/etcdmain.Main\n\t/go/src/go.etcd.io/etcd/release/etcd/server/etcdmain/main.go:40\nmain.main\n\t/go/src/go.etcd.io/etcd/release/etcd/server/main.go:32\nruntime.main\n\t/go/gos/go1.16.15/src/runtime/proc.go:225"}
```

**Comment by calvin0327** [comment]
@lonelyCZ others comments?

**Comment by lonelyCZ** [comment]
Hi, @calvin0327 , there also was an error, it not working in my env. What should I do?

```
{"level":"fatal","ts":"2023-04-11T03:36:32.566Z","caller":"etcdmain/etcd.go:204","msg":"discovery failed","error":"--initial-cluster has karmada-demo-etcd-0=http://karmada-demo-etcd-0.test.svc.cluster.local:2380 but missing from --initial-advertise-peer-urls=http://10.244.0.54:2380 (failed to resolve \"http://karmada-demo-etcd-0.test.svc.cluster.local:2380\" (lookup karmada-demo-etcd-0.test.svc.cluster.local on 10.96.0.10:53: no such host))","stacktrace":"go.etcd.io/etcd/server/v3/etcdmain.startEtcdOrProxyV2\n\t/go/src/go.etcd.io/etcd/release/etcd/server/etcdmain/etcd.go:204\ngo.etcd.io/etcd/server/v3/etcdmain.Main\n\t/go/src/go.etcd.io/etcd/release/etcd/server/etcdmain/main.go:40\nmain.main\n\t/go/src/go.etcd.io/etcd/release/etcd/server/main.go:32\nruntime.main\n\t/go/gos/go1.16.15/src/runtime/proc.go:225"}
```

**Comment by Poor12** [comment]
There are some occasional error. #3381 and #3386 is trying to track it.

**Comment by lonelyCZ** [comment]
LGTM

cc @RainbowMango , please take a look.

**Review by RainbowMango — APPROVED** [review]
/approve
Thanks.

**Pull request merged**